src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/PopupControlService.cs

```diff
@@ -236,557 +221,557 @@ private void ProcessKeyUp(object sender, KeyEventArgs e)
-        ///     This element will be replaced with the element that
-        ///     contains an active tooltip OR null if the element
-        ///     is already in play.
-        /// </param>
-        /// <param name="triggerAction">
-        ///     The user action that triggered this search.
-        /// </param>
-        /// <param name="showToolTip">
-        ///     Whether or not the tooltip found should be shown.
-        /// </param>
-        /// <returns>True if a tooltip was located.</returns>
-        private bool LocateNearestToolTip(ref DependencyObject o, ToolTip.ToolTipTrigger triggerAction, ref bool showToolTip)
+        /// <param name="o">The owner of the tooltip</param>
+        /// <param name="fromKeyboard">True if the tooltip is triggered by keyboard</param>
+        private void ShowToolTip(DependencyObject o, bool fromKeyboard)
         {
-            IInputElement element = o as IInputElement;
-            bool foundToolTip = false;
-            showToolTip = false;
+            Debug.Assert(_currentToolTip == null);
+            ResetCurrentToolTipTimer();
+            OnForceClose(null, EventArgs.Empty);
+
+            bool show = true;
 
+            IInputElement element = o as IInputElement;
             if (element != null)
             {
-                FindToolTipEventArgs args = new FindToolTipEventArgs(triggerAction);
+                ToolTipEventArgs args = new ToolTipEventArgs(opening:true);
+                // ** Public callout - re-entrancy is possible **//
                 element.RaiseEvent(args);
 
-                foundToolTip = args.Handled;
+                // [re-examine _currentToolTip, re-entrancy can change it]
+                show = !args.Handled && (_currentToolTip == null);
+            }
 
-                if (args.TargetElement != null)
+            if (show)
+            {
+                object tooltip = ToolTipService.GetToolTip(o);
+                ToolTip tip = tooltip as ToolTip;
+                if (tip != null)
+                {
+                    _currentToolTip = tip;
+                }
+                else
                 {
-                    // Open this element's ToolTip
-                    o = args.TargetElement;
-                    showToolTip =  true;
+                    _currentToolTip = new ToolTip();
+                    _currentToolTip.SetValue(ServiceOwnedProperty, BooleanBoxes.TrueBox);
+
+                    // Bind the content of the tooltip to the ToolTip attached property
+                    Binding binding = new Binding();
+                    binding.Path = new PropertyPath(ToolTipService.ToolTipProperty);
+                    binding.Mode = BindingMode.OneWay;
+                    binding.Source = o;
+                    _currentToolTip.SetBinding(ToolTip.ContentProperty, binding);
                 }
-                else if (args.KeepCurrentActive)
+
+                if (!_currentToolTip.StaysOpen)
```

Comment:
There's no harm in checking StaysOpen here, even though it will always be true in the "new ToolTip" case.  The old code did it this way, and it's good defensive programming in case someone changes the default for ToolTip.StaysOpen.